src/infer.ts

```diff
@@ -18,32 +37,122 @@ function mapObject<I, O>(obj: {[key: string]: I}, fn: ((key: string, value: I) =
+    schema_response.scalar_types[type_name_gql] = no_ops;
+    return { type: 'named', name: type_name_gql };
+  }
+
+  // OBJECT
+  // TODO: https://github.com/hasura/ndc-typescript-deno/issues/33 There should be a library function that allows us to check this case
+  else if (is_struct(ty)) {
+    const fields = Object.fromEntries(Array.from(ty.members, ([k, v]) => {
+      const field_type = checker.getTypeAtLocation(v.declarations[0].type);
+      const field_type_validated = validate_type(checker, schema_response, `${name}_field_${k}`, field_type);
+      return [k, { arguments: {}, type: field_type_validated }];
+    }));
+
+    schema_response.object_types[name] = { fields };
+    return { type: 'named', name: name}
+  }
+
+  // UNHANDLED: Assume that the type is a scalar
+  else {
+    console.error(`Unable to validate type of ${name}: ${type_str}. Assuming that it is a scalar type.`);
+    schema_response.scalar_types[name] = no_ops;
+    return { type: 'named', name };
+  }
+}
+
+export function programInfo(filename_arg?: string, vendor_arg?: string): ProgramInfo {
+  // TODO: https://github.com/hasura/ndc-typescript-deno/issues/27 This should have already been established upstream
+  const filename = resolve(filename_arg || './functions/index.ts');
+  const vendorPath = resolve(vendor_arg || './vendor');
+  const importMapPath = `${vendorPath}/import_map.json`;
   let pathsMap: {[key: string]: Array<string>} = {};
 
-  if (existsSync(vendorFilePath)) {
-    const importString = Deno.readTextFileSync(vendorFilePath);
+  if (existsSync(importMapPath)) {
+    const importString = Deno.readTextFileSync(importMapPath);
     const vendorMap = JSON.parse(importString);
     pathsMap = mapObject(vendorMap.imports, (k: string, v: string) => {
       if(/\.ts$/.test(k)) {
-        return [k, [ v.replace(/./, './vendor') ]];
+        return [k, [ v.replace(/./, vendorPath) ]];
       } else {
-        return [k.replace(/$/,'*'), [ v.replace(/./, './vendor').replace(/$/, '*') ]];
+        return [k.replace(/$/,'*'), [ v.replace(/./, vendorPath).replace(/$/, '*') ]];
       }
     });
+  } else {
+    // NOTE: We allow the import map to be optional but dependency lookup will fail if it was required.
+    console.error(`Couldn't find import map: ${importMapPath}`);
   }
 
-  const pathname = new URL('', import.meta.url).pathname;
-  const dirname = pathname.replace(/\/[^\/]*$/,'');
-  const deno_lib_path = resolve(`${dirname}/deno.d.ts`); // Assumes that deno.d.ts and infer.ts will be co-located.
+  const deno_d_ts = Deno.makeTempFileSync({ suffix: ".d.ts" });
+  Deno.writeTextFileSync(deno_d_ts, `
+  /**
+   * This module exists to be included as a library by the typescript compiler in infer.ts.
+   * The reason for this is that the user is likely to use the Deno dev tools when developing their functions.
+   * And they will have Deno in scope.
+   * This ensures that these references will typecheck correctly in infer.ts.
```

Comment:
> This ensures that these references will typecheck correctly in infer.ts.

Hrm, I wonder if this is true. Does Deno itself actually give the `any` type to its entire API, or does it actually have typings for all its functions? That could affect the the typechecker results.

For example, any function called using the below typing will return an `any` (which will then propagate to anything that uses the return value), but if the Deno tooling actually has real types, then the return type will be whatever it is, and subsequent usages of the return value will typecheck against that type.

Something to investigate later, though, not this PR.

Reply:
Yeah good point. I guess if you do some kind of utility types on something Deno returns this could be an issue.

It seems to unify simple uses of the API.

Reply:
I suspect it just _seems_ to work, but actually doesn't. `any` is basically disabling the typechecker, so I reckon you could write some code that fails to typecheck in the IDE/normal deno, but passes when run through `infer.ts`.

Good enough for now, though. For the purposes of `infer.ts`, so long as the user explicitly specifies the function return type, it probably won't affect our inference results.

Reply:
Oh yes for sure that's true. But keep in mind this is mainly for deploying already correct code, expecting that the user will have had the Deno tooling already pick up incorrect uses of Deno for them via the LSP. I think for correct Deno use, it shouldn't fail to typecheck.

Reply:
So yes, let's see if we can use the actual Deno interface in the TS compiler in future. I'll put this in the limitations docs for now.

Reply:
https://github.com/hasura/ndc-typescript-deno/issues/37